### PR TITLE
[scheduler] Improve UX for recurrence tab

### PR DESCRIPTION
**Pull request by mustafajw07**
Issue #21925

## Changelog

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

**Comment by rita-codes** [comment]
Hi @mustafajw07, thanks for picking this up! 🙌 A few things before we merge:

🔴 Read-only regression. The old `customDisabled` also checked `isPropertyReadOnly('rrule')`, but `dontRepeatDisabled` doesn't. So with rrule read-only, users can still edit interval/frequency/weekly/monthly/ends and the changes persist on save. Could you combine both? Something like:
`const inputsDisabled = controlled.recurrenceSelection === null || isPropertyReadOnly('rrule');
`

🔴 Default values lost on reopen. Pick Monthly, save, reopen: the Select shows Monthly but the fields below are empty/Weekly. The issue is in FormContent.tsx:136-137: `rruleDraft` only initializes from the saved rrule when the preset is 'custom'. In master this was hidden because fields were disabled but we want that new behavior now. Here a video showing what I mean:
https://github.com/user-attachments/assets/0e04daf9-7c4b-439e-a256-910f57db33e8

🟡 Memoization. The added useCallback/useMemo aren't paying off, the handlers are passed to MUI components that aren't React.memo-wrapped, and the option arrays aren't used as deps or passed to memoized children. Could you revert them to keep this PR focused on the UX fix?

🟡 Small stuff:
  - byMonthDay: monthlyRef.dayOfMonth ? [...] : [] — dayOfMonth is always 1–31, so the : [] branch is dead. We could simplify to [monthlyRef.dayOfMonth].
  - Rename `dontRepeatDisabled` → `inputsDisabled` (covers both reasons once you fix the read-only check).

🟢 Tests. Could you add coverage for: editing a field flipping the Select to Custom (the headline behavior, currently untested), WEEKLY/MONTHLY presets pre-filling the right values, and the two regressions above?

Thanks again! Let me know if anything's unclear 🙇‍♀️

**Comment by mustafajw07** [comment]
> Read-only regression. The old `customDisabled` also checked `isPropertyReadOnly('rrule')`, but `dontRepeatDisabled` doesn't. So with rrule read-only, users can still edit interval/frequency/weekly/monthly/ends and the changes persist on save. Could you combine both? Something like: `const inputsDisabled = controlled.recurrenceSelection === null || isPropertyReadOnly('rrule'); `
> 
> 🔴 Default values lost on reopen. Pick Monthly, save, reopen: the Select shows Monthly but the fields below are empty/Weekly. The issue is in FormContent.tsx:136-137: `rruleDraft` only initializes from the saved rrule when the preset is 'custom'. In master this was hidden because fields were disabled but we want that new behavior now. Here a video showing what I mean: https://github.com/user-attachments/assets/0e04daf9-7c4b-439e-a256-910f57db33e8
> 
> 🟡 Memoization. The added useCallback/useMemo aren't paying off, the handlers are passed to MUI components that aren't React.memo-wrapped, and the option arrays aren't used as deps or passed to memoized children. Could you revert them to keep this PR focused on the UX fix?
> 
> 🟡 Small stuff:
> 
> * byMonthDay: monthlyRef.dayOfMonth ? [...] : [] — dayOfMonth is always 1–31, so the : [] branch is dead. We could simplify to [monthlyRef.dayOfMonth].
> * Rename `dontRepeatDisabled` → `inputsDisabled` (covers both reasons once you fix the read-only check).
> 
> 🟢 Tests. Could you add coverage for: editing a field flipping the Select to Custom (the headline behavior, currently untested), WEEKLY/MONTHLY presets pre-filling the right values, and the two regressions above?

Hi @rita-codes , All the mentioned changes have been updated in the PR, including the read-only fix, default value handling, cleanup, and added test coverage. Please take a look.

File: packages/x-scheduler/src/internals/components/event-dialog/EventDialog.test.tsx [hunk 1]
```diff
@@ -1338,6 +1338,135 @@ describe('<EventDialogContent open />', () => {
...
+              eventModelStructure={rruleReadOnlyModelStructure}
+            >
+              <EventDialogContent open {...defaultProps} occurrence={recurringOccurrence} />
+            </EventCalendarProvider>,
+          );
+
+          await user.click(screen.getByRole('tab', { name: /recurrence/i }));
+
+          // Even though a recurrence is active, the detail fields must be disabled because rrule
+          // is read-only (regression: old dontRepeatDisabled only checked recurrenceSelection === null)
```
Comment (rita-codes) [inline packages/x-scheduler/src/internals/components/event-dialog/EventDialog.test.tsx]:
```suggestion
          // Even though a recurrence is active, the detail fields must be disabled because rrule is read-only
```

**Comment by rita-codes** [comment]
Hi @mustafajw07! Thanks for the updates 🙌 these changes are promising!

Found one more edge case that the new behavior opens up:

🔴 Empty byDay in WEEKLY. With the new UX you can uncheck the only selected weekday and still save. The submitted rrule ends up as {freq:'WEEKLY', byDay: []}. On reopen the Select still shows "Repeats weekly on Wednesday" (the label uses the event's start weekday, not the rule itself), but no toggle is checked because `rruleDraft.byDay` is now `[]`. Before this PR the fields were disabled so this case wasn't reachable, the new UX opens it. I've attached a video =>

https://github.com/user-attachments/assets/827f7148-ee06-462b-8b53-6f7d78a3f7d6

The simplest fix is to block unchecking the last day in `handleChangeWeeklyDays`, something like:
```
  const handleChangeWeeklyDays = (dayCode: RecurringEventWeekDayCode) => {
    setControlled((prev) => {
      const byDay = prev.rruleDraft.byDay ?? [];
      const isRemoving = byDay.includes(dayCode);
      if (isRemoving && byDay.length === 1) {
        return prev;
      }
      const next = isRemoving ? byDay.filter((d) => d !== dayCode) : [...byDay, dayCode];
      return {
        ...prev,
        recurrenceSelection: 'custom',
        rruleDraft: { ...prev.rruleDraft, byDay: next },
      };
    });
  };
```
 
What do you think? 👀 

Same issue happens via the inner frequency Select (the one inside the "Repeat every X" row). Switching to WEEKLY from there leaves byDay empty because `handleChangeFrequency` resets it. Could you pre-fill it with the event's weekday, the same way you already do for `byMonthDay` when switching to MONTHLY?
`byDay: newFrequency === 'WEEKLY' ? [monthlyRef.code] : [],`

Could you also add two small tests for these two edge cases? It would be awesome!

Thanks again for your work here!  👏

**Comment by mustafajw07** [comment]
> ame issue happens via the inner frequency Select (the one inside the "Repeat every X" row). Switching to WEEKLY from there leaves byDay empty because `handleChangeFrequency` resets it. Could you pre-fill it with the event's weekday, the same way you already do for `byMonthDay` when switching to MONTHLY?
> `byDay: newFrequency === 'WEEKLY' ? [monthlyRef.code] : [],`

Screenshot 2026-05-08 at 10 16 01 PM



How about this change — it handles the case where unchecking the last selected weekday would leave `byDay: []`, which is an invalid WEEKLY rrule. By returning `prev` unchanged, the last day stays checked and the saved rule always has at least one day. It also pre-fills `byDay` with the event's weekday when switching frequency to `WEEKLY`, matching the existing `byMonthDay` behavior for `MONTHLY`.

Let me know your thoughts.

**Comment by rita-codes** [comment]
> > ame issue happens via the inner frequency Select (the one inside the "Repeat every X" row). Switching to WEEKLY from there leaves byDay empty because `handleChangeFrequency` resets it. Could you pre-fill it with the event's weekday, the same way you already do for `byMonthDay` when switching to MONTHLY?
> > `byDay: newFrequency === 'WEEKLY' ? [monthlyRef.code] : [],`
> 
> Screenshot 2026-05-08 at 10 16 01 PM
> How about this change — it handles the case where unchecking the last selected weekday would leave `byDay: []`, which is an invalid WEEKLY rrule. By returning `prev` unchanged, the last day stays checked and the saved rule always has at least one day. It also pre-fills `byDay` with the event's weekday when switching frequency to `WEEKLY`, matching the existing `byMonthDay` behavior for `MONTHLY`.
> 
> Let me know your thoughts.

Yes, those changes look good! preventing the unchecking of the last day and pre-filling byDay on frequency switch together close the gap. And also + the small test for the unchecking case I mentioned earlier, we'll be perfectly covered! Thanks!

**Comment by mustafajw07** [comment]
> > > ame issue happens via the inner frequency Select (the one inside the "Repeat every X" row). Switching to WEEKLY from there leaves byDay empty because `handleChangeFrequency` resets it. Could you pre-fill it with the event's weekday, the same way you already do for `byMonthDay` when switching to MONTHLY?
> > > `byDay: newFrequency === 'WEEKLY' ? [monthlyRef.code] : [],`
> > 
> > 
> > Screenshot 2026-05-08 at 10 16 01 PM
> > How about this change — it handles the case where unchecking the last selected weekday would leave `byDay: []`, which is an invalid WEEKLY rrule. By returning `prev` unchanged, the last day stays checked and the saved rule always has at least one day. It also pre-fills `byDay` with the event's weekday when switching frequency to `WEEKLY`, matching the existing `byMonthDay` behavior for `MONTHLY`.
> > Let me know your thoughts.
> 
> Yes, those changes look good! preventing the unchecking of the last day and pre-filling byDay on frequency switch together close the gap. And also + the small test for the unchecking case I mentioned earlier, we'll be perfectly covered! Thanks!

Updated the pr with requested changes and also added the test cases to cover the edge cases.

**Comment by rita-codes** [comment]
All the changes look great!! ✨ 
This logic now looks so much cleaner and the extra test coverage closes the gaps cleanly.

Thanks so much for the great work here, this is a really nice UX improvement for the recurrence tab! 🙌

Approving! :shipit:

**Review by rita-codes — APPROVED** [review]

**Comment by mustafajw07** [comment]
> All the changes look great!! ✨ This logic now looks so much cleaner and the extra test coverage closes the gaps cleanly.
> 
> Thanks so much for the great work here, this is a really nice UX improvement for the recurrence tab! 🙌
> 
> Approving! :shipit:

Really appreciate the feedback! 
Glad the cleanup and test coverage improvements helped improve the overall UX experience 🙌
